Raven db 10953

talweiss1982 opened the pull request.

Commit:
RavenDB-10953 Replicating missing attachments

Commit:
RavenDB-10953 removing hacks to debug assertion

arekpalinski commented on src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:
```diff
@@ -341,11 +348,31 @@ void HandleSocketException(SocketException e)
 
+                DynamicJsonValue returnValue;
+
+                if (e is AggregateException ag && ag.InnerExceptions.Count == 1 && ag.InnerExceptions[0] is MissingAttachmentException mae)
```
You can use `ExtractSingleInnerException` to simplify that

talweiss1982 replied:
okay

arekpalinski commented on src/Raven.Server/Documents/AttachmentsStorage.cs:
```diff
@@ -467,6 +469,30 @@ public IEnumerable<Attachment> GetAttachmentsForDocument(DocumentsOperationConte
+            }
+        }
+
+        private static ByteStringContext.InternalScope From(DocumentsOperationContext context, LazyStringValue lowerDocumentId, out Slice lowerDocumentIdSlice)
```
Either rename this method or drop it and call `Slice.From` directly

talweiss1982 replied:
its unsafe method and you can't call unsafe from IEnumerable

arekpalinski replied:
ok, makes sense

arekpalinski commented on src/Raven.Server/Documents/AttachmentsStorage.cs:
```diff
@@ -558,6 +589,11 @@ public Stream GetAttachmentStream(DocumentsOperationContext context, Slice hashS
             return tree.ReadStream(hashSlice);
         }
 
+        public bool VerifyAttachmentStreamExist(DocumentsOperationContext context, Slice hashSlice)
```
Let's make it private

talweiss1982 replied:
okay

arekpalinski commented on src/Raven.Server/Documents/AttachmentsStorage.cs:
```diff
@@ -522,12 +548,17 @@ public Attachment GetAttachment(DocumentsOperationContext context, string docume
+            // We were invoked just for the purpose of checking the existence of the attachment, no need to actually read the stream from disk only to verify its there.
+            if (withoutStream)
+            {
+                return VerifyAttachmentStreamExist(context, attachment.Base64Hash)?attachment:null;
```
Why is it possible to have an attachment without the stream?

arekpalinski replied:
BTW it isn't obvious for the caller of this method that `withoutStream` parameter changes the logic and calls  `VerifyAttachmentStreamExist` - it looks the storage should have a dedicated method for that check

talweiss1982 replied:
This is an optimization so we won't need to fetch the stream, it is invoked just to verify the existence of a stream without having to read the stream from the disk.

ayende replied:
Since we just want to verify existence, why not just have an `AttachmentExists` method?

ayende replied:
I agree with @arekpalinski that the boolean arg changing the meaning so drastically is better done with a different method entirely.

talweiss1982 replied:
i have not changed the logic of the method even by a bit, was just trying to save the actual read of the stream from the disk where only its existence i care about.
I see that effort is not been appreciated next time I'll just invoke code like /databases/*/attachments HEAD

arekpalinski commented on src/Raven.Server/Documents/AttachmentsStorage.cs:
```diff
@@ -997,5 +1033,26 @@ public static void AssertAttachments(BlittableJsonReaderObject document, Documen
 #endif
+        public void AssertAttachmentsFromReplication(BlittableJsonReaderObject document, AttachmentType type,string id, DocumentsOperationContext context)
+        {
+            var changeVector = type == AttachmentType.Document ? null : document.GetChangeVector();
```
Is it supposed to work for attachments of revisions as well (`AttachmentType.Revision`)?

talweiss1982 replied:
it suppose to unless you know something i don't, what is the issue here?

arekpalinski replied:
Just asking. I wondered why we pass `null` here for revisions - I saw your other comment about the key that we search on and understand now

arekpalinski commented on src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:
```diff
@@ -525,7 +554,17 @@ private unsafe void ReceiveSingleDocumentsBatch(DocumentsOperationContext docume
-                    _log.Info("Failed to receive documents replication batch. This is not supposed to happen, and is likely a bug.", e);
+                {
+                    //This is the case where we had a missing attachment, it is rare but expected.
+                    if (e is AggregateException ag && ag.InnerExceptions.Count == 1 && ag.InnerExceptions[0] is MissingAttachmentException mae)
```
Please use `ExtractSingleInnerException `

talweiss1982 replied:
okay

arekpalinski commented on src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs:
```diff
@@ -240,6 +244,22 @@ public bool ExecuteReplicationOnce(OutgoingReplicationStatsScope stats, ref Date
+                            {
+                                var type = (item.Flags & DocumentFlags.Revision) == DocumentFlags.Revision ? AttachmentType.Revision: AttachmentType.Document;
+                                foreach (var attachment in _parent._database.DocumentsStorage.AttachmentsStorage.GetAttachmentsForDocument(documentsContext, type, item.Id)
+                                    .Where(a => a.Etag <= firstBatchEtag).Select(a => ReplicationBatchItem.From(a)))
```
We can avoid Linq here and filtering it explicitly in the loop - it might be slightly more optimized way

talweiss1982 replied:
this is a rare case and will only invoke if there are missing attachments and if this document has attachments so the performance penalty is neglectable.

ayende replied:
I don't think we need to do any of that. 
You already _have_ all the attachments you need, in the metadata of the document, no?

talweiss1982 replied:
no

ayende replied:
why not?

ayende replied:
The reason we require it here is that we need to send all the attachment's data.
That include the change vector, content-type, etc. These are stored on a per attachment, not at the hash level.

ayende replied:
However, we don't need to have the `Where` here, we can rely on the fact that attachments are _always_ modified before the document.

ayende replied:
Therefor, we can remove all the `Linq` in here completely

arekpalinski commented on src/Voron/Data/BTrees/Tree.Stream.cs:
```diff
@@ -257,6 +257,27 @@ public VoronStream ReadStream(Slice key)
+        /// </summary>
+        /// <param name="key"> The key of the stream</param>
+        /// <returns>'true' if there is a stream under the given 'key' and 'false' otherwise</returns>
+        public bool SeekStream(Slice key)
```
As the purpose of this method is to check the existence  of the stream we'll be better of by renaming it to `Exists` or something like that

talweiss1982 replied:
okay

arekpalinski commented:
Please check failing replication tests

Commit:
RavenDB-10953 PR comments

Commit:
RavenDB-10953 fixing failing tests

ayende commented on src/Raven.Server/Documents/AttachmentsStorage.cs:
```diff
@@ -997,5 +1033,26 @@ public static void AssertAttachments(BlittableJsonReaderObject document, Documen
+                    if (attachment.TryGet(nameof(AttachmentName.Name), out LazyStringValue name) &&
+                        attachment.TryGet(nameof(AttachmentName.Hash), out LazyStringValue hash))
+                    {
+                        if (GetAttachment(context, id, name, type, changeVector, withoutStream:true) == null)
```
Here you don't need any of the validation that happens in `GetAttachment`, can probably call `VerifyAttachmentStreamExist` directly

ayende commented on src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:
```diff
@@ -341,11 +348,31 @@ void HandleSocketException(SocketException e)
+
+                if (e.ExtractSingleInnerException() is MissingAttachmentException mae)
+                {
+                    if (_cts.IsCancellationRequested)
```
We just checked this, no need to do so again

ayende commented on src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:
```diff
@@ -1080,6 +1119,8 @@ public override int Execute(DocumentsOperationContext context)
                                         //the other side will receive negative ack and will retry sending again.
                                         document = new BlittableJsonReaderObject(_buffer + item.Position, item.DocumentSize, context);
                                         document.BlittableValidation();
+                                        var attachmentType = item.Flags.Contain(DocumentFlags.Revision) ? AttachmentType.Revision : AttachmentType.Document;                                        
```
Pretty sure that we don't really care if this is revision or document.
We only care if the attachment hash exists

talweiss1982 replied:
The type effects the key we search for and it is importent, so we do care.

ayende replied:
Why do you need a type? All you need to do is check if the hash is there, and you already have this from the metadata.

talweiss1982 replied:
the type defines if the key of the attachment is constructed by the document id or it's change vector if it is a revision the change vector replace the id, why? I don't know I haven't written it, this is whats implemented.

ayende commented on src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs:
```diff
@@ -300,10 +322,17 @@ public bool ExecuteReplicationOnce(OutgoingReplicationStatsScope stats, ref Date
                         throw;
                     }
+
+                    if (hadMissingAttachment)
```
why not just set unconditionally

talweiss1982 replied:
Code refactoring, leftovers...

ayende commented on src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs:
```diff
@@ -441,7 +470,8 @@ private bool AddReplicationItemToBatch(ReplicationBatchItem item, OutgoingReplic
 
             // destination already has it
-            if (ChangeVectorUtils.GetConflictStatus(item.ChangeVector, _parent.LastAcceptedChangeVector) == ConflictStatus.AlreadyMerged)
+            if (MissingAttachmentsInLastBatch == false && 
```
When will this happen?

talweiss1982 replied:
This is the code that filters items that have lower CV, this will filter the missing attachments if we ignore MissingAttachmentsInLastBatch == false

ayende commented on src/Voron/Data/BTrees/Tree.Stream.cs:
```diff
@@ -257,6 +257,27 @@ public VoronStream ReadStream(Slice key)
+            if (numberOfChunks <= 0)
+                return false;
+
+            using (var it = tree.Iterate())
```
Why? This must be true, you already checked `NumberOfEntries`

talweiss1982 replied:
I was not familiar with the code enough to realize reading the Tree is enough

Commit:
RavenDB-10953

talweiss1982 commented:
var type = (item.Flags & DocumentFlags.Revision) == DocumentFlags.Revision ? AttachmentType.Revision: AttachmentType.Document;
                                foreach (var attachment in _parent._database.DocumentsStorage.AttachmentsStorage.GetAttachmentsForDocument(documentsContext, type, item.Id)
                                    .Where(a => a.Etag <= firstBatchEtag).Select(a => ReplicationBatchItem.From(a)))

This is the code that actually brings the attachments into the batch why don't we need it ?
where do expect me to fetch the attachments from?
all we have are the descriptor of attachments in the document's metadata not the actual attachments

ppekrol commented:
@talweiss1982 can you check failing tests?

talweiss1982 commented:
@ppekrol the majority of them are failing because of conflicts not been created (not a bug but a feature) need to go over them one by one and verify whether a bug or a feature

ppekrol commented:
Ok, great.

Commit:
RavenDB-10953 Comments fix

ayende commented on src/Raven.Server/Documents/AttachmentsStorage.cs:
```diff
@@ -467,6 +470,30 @@ public IEnumerable<Attachment> GetAttachmentsForDocument(DocumentsOperationConte
+            }
+        }
+
+        private static ByteStringContext.InternalScope GetSliceFromLazyString(DocumentsOperationContext context, LazyStringValue lowerDocumentId, out Slice lowerDocumentIdSlice)
```
Not sure if this is worth a separate method.

ayende replied:
It isn't worth it, but this is required by the compiler, cannot use `yield` and `unsafe` together.

ayende replied:
Fixed this by adding a new `Slice.From` overload

ayende commented on src/Raven.Server/Documents/AttachmentsStorage.cs:
```diff
@@ -997,5 +1030,25 @@ public static void AssertAttachments(BlittableJsonReaderObject document, Documen
+        {
+            if (document.TryGet(Constants.Documents.Metadata.Key, out BlittableJsonReaderObject metadata) &&
+                metadata.TryGet(Constants.Documents.Metadata.Attachments, out BlittableJsonReaderArray attachments))
+            {
```
Let's get everything about the document itself out of here, we shouldn't have the document's metadata details in attachment storage

ayende committed:
RavenDB-10953
- Add overload for Slice.From to make it easier to use LazyStringValue slices from methods that yield
- Avoid using Linq when sending attachments when MissingAttachmentsInLastBatch is set
- Moved asserting that attachment's hashes are found to the incoming replication, instead of attachment storage

ayende committed:
RavenDB-10953 Avoid sending already seen values to the other side (unless they are attachments and we are in MissingAttachmentsInLastBatch mode)

ayende commented:
Pushed the wrong branch, forcing pushing.
